godot-core/src/builtin/packed_array.rs

```diff
@@ -0,0 +1,544 @@
+            fn ptr(&self, index: usize) -> *const $Element {
+                self.check_bounds(index);
+                // SAFETY: We just checked that the index is not out of bounds.
+                let ptr = unsafe {
+                    (interface_fn!($operator_index_const))(self.sys(), to_i64(index)) as *const $Element
+                };
+                assert!(!ptr.is_null());
+                ptr
+            }
+
+            /// Returns a mutable pointer to the element at the given index.
+            ///
+            /// # Panics
+            ///
+            /// If `index` is out of bounds.
+            fn ptr_mut(&self, index: usize) -> *mut $Element {
+                self.check_bounds(index);
+                // SAFETY: We just checked that the index is not out of bounds.
+                let ptr = unsafe {
+                    (interface_fn!($operator_index))(self.sys(), to_i64(index)) as *mut $Element
+                };
+                assert!(!ptr.is_null());
+                ptr
+            }
+
+            /// Converts an `$Element` into a value that can be passed into API functions. For most
+            /// types, this is a no-op. But `u8` and `i32` are widened to `i64`, and `f32` is
+            /// widened to `f64`.
+            #[inline]
+            fn into_arg(e: $Element) -> $Arg {
+                e.into()
+            }
+
+            #[doc(hidden)]
+            pub fn as_inner(&self) -> inner::$Inner<'_> {
+                inner::$Inner::from_outer(self)
+            }
+        }
+
+        impl_builtin_traits! {
+            for $PackedArray {
+                Default => $construct_default;
+                Clone => $construct_copy;
+                Drop => $destroy;
+            }
+        }
+
+        /// Creates a `$PackedArray` from the given Rust array.
+        #[cfg(not(any(gdext_test, doctest)))]
+        impl<const N: usize> From<&[$Element; N]> for $PackedArray {
+            fn from(arr: &[$Element; N]) -> Self {
+                Self::from(&arr[..])
+            }
+        }
+
+        /// Creates a `$PackedArray` from the given slice.
+        #[cfg(not(any(gdext_test, doctest)))]
+        impl From<&[$Element]> for $PackedArray {
+            fn from(slice: &[$Element]) -> Self {
+                let mut array = Self::new();
+                let len = slice.len();
+                if len == 0 {
+                    return array;
+                }
+                array.resize(len);
+                let ptr = array.ptr_mut(0);
+                for (i, element) in slice.iter().enumerate() {
+                    // SAFETY: The array contains exactly `len` elements, stored contiguously in memory.
+                    unsafe {
+                        // `GodotString` does not implement `Copy` so we have to call `.clone()`
+                        // here.
+                        *ptr.offset(to_isize(i)) = element.clone();
+                    }
+                }
+                array
```

Comment:
Could this not be moved to `FromIterator` and then reused from here?

There, instead of `slice.len()` we could use `size_hint()`, I'm almost certain that provides an accurate number for slices.

Reply:
It does for slices, but not for iterators in general. So we would need a lot more complexity to deal with iterators that don't have a size hint, whose size hint is incorrect, or even whose size hint changes during iteration. Let's keep it simple and stupid for now.